[Merged by Bors] - Implement demo command (#66)
stackabletech/stackablectl: docs/modules/ROOT/pages/demos/trino-taxi-data.adoc

= trino-taxi-data

This demo will

> * Install the needed Stackable operators

Reviewer: ```suggestion
* Install the required Stackable operators
```

* Spin up the follow data products
** Superset: A modern data exploration and visualization platform. This demo uses it to execute SQL queries and build dashboards
** Trino: A fast distributed SQL query engine for big data analytics that helps you explore your data universe. This demo uses it to enable SQL access to the data

> ** MinIO: A S3 compatible object store. This demo uses it as long-term storage.

Reviewer: Not sure about this one :D
```suggestion
** MinIO: A S3 compatible object store. This demo uses it as persistent storage.
```

Author: Changed to `** MinIO: A S3 compatible object store. This demo uses it as a persistent storage to store all the data used`

** Hive metastore: A service that stores metadata related to Apache Hive and other services. This demo uses it as metadata storage for Trino

> * Load testdata into S3. It contains 2 and a half years of New York City taxi trips.

Reviewer: I would go for either both text or both numbers
```suggestion
* Load testdata into S3. It contains 2.5 years of New York City taxi trips.
```

* Make data accessible via SQL in Trino
* Create Superset dashboards for visualization

You can see the deployed products as well as their relationship in the following diagram:

image::demo-trino-taxi-data/overview.png[]

== Inspect data in S3

> The S3 provided by MinIO provides long-term storage.

Reviewer: If you adapt the persistent above, keep it consistent here. I never really heard of long-term storage.

Author: Changed to `The S3 provided by MinIO is used as persistent storage to store all the data used`

=== Inspect dashboard
On the top click on the tab `Dashboards`.

image::demo-trino-taxi-data/superset_3.png[]

Click on the dashboard called `Taxi data`.
It might take some time until the dashboards renders all the included charts.

image::demo-trino-taxi-data/superset_4.png[]

You can clearly see the impact of Covid-19 on the taxi business.

=== Execute arbitrary SQL statements
Within Superset you can not only create dashboards but also run arbitrary SQL statements.
On the top click on the tab `SQL Lab` -> `SQL Editor`.

image::demo-trino-taxi-data/superset_5.png[]

On the left select the database `Trino`, the schema `demo` and set `See table schema` to `ny_taxi_data`.

image::demo-trino-taxi-data/superset_6.png[]

On the right textbox enter the desired SQL statement.
If you do not want to make on up you can use the following:

[source,sql]
----
select
  format_datetime(tpep_pickup_datetime, 'YYYY/MM') as month,
  count(*) as trips,
  sum(total_amount) as sales,
  avg(date_diff('minute', tpep_pickup_datetime, tpep_dropoff_datetime)) as avg_duration_min
from ny_taxi_data
group by 1
order by 1
----

image::demo-trino-taxi-data/superset_7.png[]

== Summary

> The demo loaded 2 and a half years of taxi trip data from New York City with 68 million records and a total size of 1GB in parquet files.

Reviewer: ```suggestion
The demo loaded 2.5 years of taxi trip data from New York City with 68 million records and a total size of 1GB in parquet files.
```

             3.0 |   2504112 
             2.0 |   9575299 
             1.0 |  48133494 
             0.0 |   1454268 
            NULL |   2698591 
----

We can see that one trip had 112 and another one 96 passengers.
All the other trips start with a more "realistic" number of 9 passengers.

As a bonus question: What *exactly* did the large passenger do?

[source,sql]
----
select *
from ny_taxi_data
where passenger_count > 50
----

returns

[source]
----
 vendorid |  tpep_pickup_datetime   |  tpep_dropoff_datetime  | duration_min | passenger_count | trip_distance | payment_type | fare_amount | tip_amount | total_amount 
----------+-------------------------+-------------------------+--------------+-----------------+---------------+--------------+-------------+------------+--------------
        2 | 2021-08-01 19:47:43.000 | 2021-08-01 19:57:54.000 |           10 |           112.0 |           1.8 | Credit card  |         9.0 |       2.46 |        14.76 
        2 | 2021-08-03 11:51:58.000 | 2021-08-03 12:09:29.000 |           17 |            96.0 |          1.56 | Credit card  |        11.5 |       2.22 |        17.02 
----
Pretty cheap for that amount of persons!

> This probably is a invalid record.

Reviewer: ```suggestion
This probably is an invalid record.
```
